Add GitLab code quality formatter

mschoettle opened the pull request:
Add a formatter that outputs a GitLab code quality report artifact JSON file.

As suggested I left the author field pointing to me. I also added a `LICENSE` with your name since it will be released under your name. Let me know if this doesn't make sense.

I also added a README and tried to take care of all the tests.

## Open issues/questions

- [x] running `npm test` produces the output file in `test/formatter-gitlab/` which is not automatically deleted. How can this be achieved?
- [x] looking at the diff of the snapshots it seems that the JSON output shows up in the wrong place (within `formatterJson`). Is this because the formatter is missing from `test/markdownlint-cli2-test-cases.js` lines 62 and 71?
- [x] does it need to be added to `test/outputFormatters-pre-imported/.markdownlint-cli2.cjs` as well?
- [x] I would like to have a test case that ensures that the fingerprint is different for the same violation appearing multiple times in the same file on different lines (I created this scenario in `test/formatter-gitlab/dir/subdir/info.md`. What would be the best way to add a test case for this?
- [x] the spec is a subset of the codeclimate spec. In theory it would not be very difficult to support the full spec. Maybe naming the formatter `-codeclimate` would be better to prepare for it (or support it right away)?

## TODO

- [x] squash commits into one and reference the issue

add `markdownlint-cli2-formatter-gitlab` to (I assume it would require the package to be published first in most cases)

- [ ] `test/outputFormatters-npm/.markdownlint-cli2.jsonc`
- [ ] `package.json` `devDependencies`
- [ ] `dictionary.txt`
- [ ] `.github/workflows/checkers.yml`
- [ ] `.github/workflows/ci.yml`

Fixes #92

mschoettle commented:
An unrelated question: Would it be possible to include (install) the formatters in the Docker image so the image can be used without having to install them?

DavidAnson commented:
Thank you, this looks very thorough! I will provide initial feedback later today, but may not be able to do a full review for a little while. Thanks for your patience! Regarding including formatters in the Docker container image, that should already be true. I have separate tests for that, but I see they don't explicitly check formatters, so it's possible that's broken?

https://github.com/DavidAnson/Workflows/blob/main/.github/workflows/smoke-test.yml

mschoettle commented:
> Regarding including formatters in the Docker container image, that should already be true. I have separate tests for that, but I see they don't explicitly check formatters, so it's possible that's broken?

I had a quick look inside the image and I could only find the default formatter installed. I could have a look into this if all the formatters should be installed in the image.

DavidAnson commented:
Please have a look at the Dockerfile, I think they should get installed globally. That's why I use asterisk there. It's possible the publish steps changed slightly such that they are no longer present. Ah, this is possibly related to the Docker publish step now happening in a GitHub Action instead of locally, which is how I originally had things working.

DavidAnson commented:
From what I'm seeing on my phone, I agree it looks broken. I've made myself a note to sort that out separately.

mschoettle commented:
> Please have a look at the Dockerfile, I think they should get installed globally. That's why I use asterisk there.

From my understanding it installs only the non-dev dependencies of `markdownlint-cli2` which only includes the default formatter. The other formatters are listed as dev dependencies. It might require to install them specifically in the `Dockerfile` to avoid installing all formatters in other scenarios.

DavidAnson commented:
At one point, I think all the pack files got created locally, and that command installed them all globally into the image. Again, I agree that doesn't seem likely anymore. I will have a look.

DavidAnson commented:
> running npm test produces the output file in test/formatter-gitlab/ which is not automatically deleted. How can this be achieved?

There's some kind of wrapper in the tests that copies to a new directory and cleans up after it, I think. It's been a while since I've looked at this, and I am doing so now from my phone, so some of the specifics are a little hazy.

> does it need to be added to test/outputFormatters-pre-imported/.markdownlint-cli2.cjs as well?

Ideally, this new formatters would show up in all the places where the other formatters are grouped.

> I would like to have a test case that ensures that the fingerprint is different for the same violation appearing multiple times in the same file on different lines (I created this scenario in test/formatter-gitlab/dir/subdir/info.md. What would be the best way to add a test case for this?

If you end up doing a diff of output like I think the JSON formatter tests do, this should be handled implicitly if your test cases include the scenario you care about.

> the spec is a subset of the codeclimate spec. In theory it would not be very difficult to support the full spec. Maybe naming the formatter -codeclimate would be better to prepare for it (or support it right away)?

Yes, if this would be useful in scenarios outside GitLab, that might be a good way to clarify that. (I'm not familiar with this part of the test ecosystem.)

> add markdownlint-cli2-formatter-gitlab to (I assume it would require the package to be published first in most cases)

Yes, that's a follow-up commit after publish.

DavidAnson commented on formatter-gitlab/README.md:
```diff
@@ -0,0 +1,121 @@
+
+> An output formatter for [`markdownlint-cli2`][markdownlint-cli2] that writes
+> results to a [GitLab code quality report artifact][gitlab] [JSON][json] file
+> (a subset of the [CodeClimate spec][codeclimate])
```
Spell out "specification", please

DavidAnson commented on formatter-gitlab/README.md:
```diff
@@ -0,0 +1,121 @@
+
+## Use
+
+For the default output file name of `"markdownlint-cli2-results.json"`, use
```
This should be unique for this formatter, either by extension or by name.

DavidAnson commented on formatter-gitlab/README.md:
````diff
@@ -0,0 +1,121 @@
+}
+```
+
+To customize the output file name or number of spaces to indent, use the
````
Is indent a meaningful parameter here? I am thinking it should not be because the purpose of this file is to be ingested by some other tool. It was only included with the JSON formatter because some people might have used that to review visually.

DavidAnson replied:
Same comment for output file name. If folks wouldn't need to customize it, let's leave it out.

mschoettle replied:
Good point. I removed both of them.

DavidAnson commented on formatter-gitlab/markdownlint-cli2-formatter-gitlab.js:
```diff
@@ -0,0 +1,76 @@
+
+const fs = require("node:fs").promises;
+const path = require("node:path");
+
```
Remove blank line

DavidAnson commented on formatter-gitlab/markdownlint-cli2-formatter-gitlab.js:
```diff
@@ -0,0 +1,76 @@
+const { createHash } = require("node:crypto");
+
+/**
+ * @param {string} violation The complete textual description of the violation
```
I think my pattern is to end all these with a period?

DavidAnson commented on formatter-gitlab/markdownlint-cli2-formatter-gitlab.js:
```diff
@@ -0,0 +1,76 @@
+    const description = ruleDescription +
+          (errorDetail ? ` [${errorDetail}]` : "") +
+          (errorContext ? ` [Context: "${errorContext}"]` : "");
+    // construct error text with all details to use for unique fingerprint
```
Is this your technique, or is this part of the specifications? Also, I think comments should begin with a capital letter.

mschoettle replied:
This is to create a unique fingerprint for the violation. I reused the code from the default formatter to produce the text.

The [code climate specification](https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#data-types) states for the `fingerprint`:

> A unique, deterministic identifier for the specific issue being reported to allow a user to exclude it from future analyses.

DavidAnson commented on formatter-gitlab/markdownlint-cli2-formatter-gitlab.js:
```diff
@@ -0,0 +1,76 @@
+  const content = JSON.stringify(issues, null, spaces || 2);
+  return fs.writeFile(
+    path.resolve(
+      // eslint-disable-next-line no-inline-comments
```
I want to get rid of this, unless I also used it in the JSON formatter. :)

mschoettle replied:
You did :). I removed it and it seems to work still.

DavidAnson commented on test/markdownlint-cli2-test-cases.js:
```diff
@@ -689,6 +689,12 @@ const testCases =
     "usesEnv": true
   });
 
+  testCase({
```
Look at pre/post copy/delete directory from other tests.

mschoettle replied:
I removed it from here (I think this is was caused the file). The JSON and XML formatter are not listed as a testcase here either.

DavidAnson reviewed (changes requested):
There is quite a lot here! Thank you for your attention to detail! I left a first pass of comments and tried to answer some of your questions. I didn't see anything nothing significant/concerning.

mschoettle commented:
I addressed your comments. 

* changed formatter name to `formatter-codequality`
* changed output file to be unique
* removed all parameters from the formatter
* I was able to update the snapshots, had to manually add `formatterCodeQuality: ''` to a lot of those and then run `npx ava --update-snapshots`. Hopefully that was the correct procedure.

I tried to output the JSON onto `stdout` instead of writing to a file. This would allow it to be invoked and the output directly piped to some other tool/file. But the extra output (version and number of files and errors) prevents this.

DavidAnson commented on formatter-codequality/package.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "markdownlint-cli2-formatter-codequality",
+  "version": "0.0.1",
+  "description": "An output formatter for markdownlint-cli2 that writes results to a JSON file following the GitLab code quality syntax",
```
Please make this the same as the initial blockquote in README (okay to leave out the parenthesized part at the end).

DavidAnson commented on test/snapshots/markdownlint-cli2-test-fs.js.md:
```diff
@@ -10,6 +10,7 @@ Generated by [AVA](https://avajs.dev).
 
     {
       exitCode: 1,
+      formatterCodeQuality: '',
```
LOL, sorry...

DavidAnson commented on formatter-codequality/markdownlint-cli2-formatter-codequality.js:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * @param {string} violation The complete textual description of the violation.
+ * @returns {string} The SHA256 fingerprint for the violation.
```
The SHA256 fingerprint for the violation as a hexadecimal string.

DavidAnson commented on formatter-codequality/markdownlint-cli2-formatter-codequality.js:
```diff
@@ -0,0 +1,73 @@
+};
+
+// Writes markdownlint-cli2 results to a
+// GitLab code quality report artifact JSON file
```
Nit: Merge with previous line?

DavidAnson commented on formatter-codequality/markdownlint-cli2-formatter-codequality.js:
```diff
@@ -0,0 +1,73 @@
+// Writes markdownlint-cli2 results to a
+// GitLab code quality report artifact JSON file
+// eslint-disable-next-line max-len
+// see: https://docs.gitlab.com/ee/ci/testing/code_quality.html#implementing-a-custom-tool
```
Nit: "See"

DavidAnson commented on formatter-codequality/markdownlint-cli2-formatter-codequality.js:
```diff
@@ -0,0 +1,73 @@
+// see: https://docs.gitlab.com/ee/ci/testing/code_quality.html#implementing-a-custom-tool
+const outputFormatter = (options) => {
+  const { directory, results } = options;
+
```
Nit: Remove blank line

DavidAnson reviewed (changes requested):
This looks awesome! I left a few very minor comments. Should be easy to fix or let me know if you're tired of this and I'll do so. :)

mschoettle commented:
> This looks awesome! I left a few very minor comments. Should be easy to fix or let me know if you're tired of this and I'll do so. :)

Not a problem at all!

Fixed all of them and also added a snippet to the README on how to use this in a GitLab CI job.

DavidAnson commented:
I couldn't push changes to your branch from Codespaces, so I made the last few tweaks in my own repo. Your commit is now present in `next` branch as https://github.com/DavidAnson/markdownlint-cli2/commit/50855cd35f50a3d5336279e3b9d065c0df3251f5.

Thank you very much!

This part took longer than I planned tonight, so I will publish the package to npm tomorrow and make the remaining updates to this repo to test the published package.

mschoettle commented:
Great, thanks a lot!
